posthog/api/plugin.py

```diff
@@ -0,0 +1,132 @@
+from rest_framework.response import Response
+
+from posthog.models import Plugin, PluginConfig
+from posthog.plugins import download_plugin_github_zip, reload_plugins_on_workers
+
+
+class PluginSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Plugin
+        fields = ["id", "name", "description", "url", "config_schema", "tag", "from_json"]
+
+    def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> Plugin:
+        if not settings.PLUGINS_INSTALL_FROM_WEB:
+            raise APIException("Plugin installation via the web is disabled!")
+        if len(Plugin.objects.filter(name=validated_data["name"])) > 0:
+            raise APIException('Plugin with name "{}" already installed!'.format(validated_data["name"]))
+        validated_data["archive"] = download_plugin_github_zip(validated_data["url"], validated_data["tag"])
+        if "from_json" in validated_data:  # prevent hackery
+            del validated_data["from_json"]
+        plugin = Plugin.objects.create(from_web=True, **validated_data)
+        reload_plugins_on_workers()
+        return plugin
+
+    def update(self, plugin: Plugin, validated_data: Dict, *args: Any, **kwargs: Any) -> Plugin:  # type: ignore
+        if not settings.PLUGINS_INSTALL_FROM_WEB:
+            raise APIException("Plugin installation via the web is disabled!")
+        if plugin.from_json:
+            raise APIException('Can not update plugin "{}", which is configured from posthog.json!'.format(plugin.name))
+        plugin.name = validated_data.get("name", plugin.name)
+        plugin.description = validated_data.get("description", plugin.description)
+        plugin.url = validated_data.get("url", plugin.url)
+        plugin.config_schema = validated_data.get("config_schema", plugin.config_schema)
+        plugin.tag = validated_data.get("tag", plugin.tag)
+        plugin.archive = download_plugin_github_zip(plugin.url, plugin.tag)
+        plugin.save()
+        reload_plugins_on_workers()
+        return plugin
+
+
+class PluginViewSet(viewsets.ModelViewSet):
+    queryset = Plugin.objects.all()
+    serializer_class = PluginSerializer
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        if not settings.PLUGINS_INSTALL_FROM_WEB and not settings.PLUGINS_CONFIGURE_FROM_WEB:
+            return queryset.none()
+        return queryset
+
+    @action(methods=["GET"], detail=False)
+    def repository(self, request: request.Request):
+        if not settings.PLUGINS_INSTALL_FROM_WEB:
+            return Response([])
+        url = "https://raw.githubusercontent.com/PostHog/plugins/main/repository.json"
+        plugins = requests.get(url)
+        return Response(json.loads(plugins.text))
+
+    def destroy(self, request: request.Request, pk=None) -> Response:  # type: ignore
+        if not settings.PLUGINS_INSTALL_FROM_WEB:
+            raise APIException("Plugin installation via the web is disabled!")
```

Comment:
Seeing this I think it might be better to do this validation in a central place. Perhaps it can be a `Permission` ?

Reply:
I'd definitely like to revisit this logic! The idea though is to get something quickly out.
In the future I believe plugins should be either enabled or disabled via the global `PLUGINS_ENABLED` setting... and more detailed access for installing/configuring would be given to the equivalent of "superadmins" (instance admins?) and project admins.

I would like to visit this permission issue together with integration into the new organization structure in a separate PR though.